rmw_implementation/src/functions.cpp

```diff
@@ -534,6 +534,24 @@ RMW_INTERFACE_FN(rmw_set_log_severity,
   rmw_ret_t, RMW_RET_ERROR,
   1, ARG_TYPES(rmw_log_severity_t))
 
+RMW_INTERFACE_FN(rmw_get_publishers_info_by_topic,
+  rmw_ret_t, RMW_RET_UNSUPPORTED,
+  5, ARG_TYPES(
+    const rmw_node_t *,
+    rcutils_allocator_t *,
+    const char *,
+    bool,
+    rmw_topic_info_array_t *))
+
+RMW_INTERFACE_FN(rmw_get_subscriptions_info_by_topic,
+  rmw_ret_t, RMW_RET_UNSUPPORTED,
```

Comment:
Returning `RMW_RET_UNSUPPORTED` as I think this is more apt than `RMW_RET_ERROR`

Reply:
We usually return `RMW_RET_ERROR` in this case.
In the case a specific rmw implementation doesn't support the feature, it has to still implement a function just returning `RMW_RET_UNSUPPORTED`.

Reply:
Hmm, maybe this implementation needs to be modified to catch `UNSUPPORTED` functions and return `RMW_RET_UNSUPPORTED` for those since this is an error caught [here](https://github.com/ros2/rmw_implementation/blob/6dd0abf8380378c6dc2e432aed2a73f049db9856/rmw_implementation/src/functions.cpp#L159) anyway and return `RMW_RET_ERROR` for the others. 

This is particularly important in throwing specific errors on the client layers (like [this](https://github.com/ros2/rclpy/pull/454/files#diff-80dda03110b5be606b823d6b0558b5c2R889)).

I guess, for now I'll change this back to `RMW_RET_ERROR`

Reply:
The problem, is that there are two ways of linking against `rmw`.

If there is more than one implementation, the code here in `rmw_implementation` is used, with all the dynamic loading stuff and symbol finding magic.

If you only have one implementation, the code is directly linked to the specific rmw implementation, ignoring `RMW_IMPLEMENTATION` environment variable at runtime (check https://github.com/ros2/rmw_implementation/blob/6dd0abf8380378c6dc2e432aed2a73f049db9856/rmw_implementation/CMakeLists.txt and https://github.com/ros2/rmw_implementation/blob/6dd0abf8380378c6dc2e432aed2a73f049db9856/rmw_implementation/rmw_implementation-extras.cmake.in).

If a specific rmw implementation doesn't implement a function, in the first case is detected at runtime and reported with an error code, and in the second linking would directly fail.

`RMW_RET_UNSUPPORTED` means that the specific rmw implementation that you're using doesn't support that feature, and that you can't use it as an user. It's not the idea to return that error if the symbol is not found, because that doesn't mean that the rmw implementation doesn't support that feature, but that is wrongly implemented and someone has to report a bug. IMO, `RMW_RET_ERROR` is the correct error code (and also consistent with the existing  code).

In any case, this is out of the scope of the PR. Please, open a follow-up issue if you think that this should be handled differently.